Replace custom ChaCha20 cipher with RustCrypto ChaCha20 implementation

philipr-za opened the pull request:
## Description
Fuzzing revealed that the custom ChaCha20 implementation in `tari_utilities` was not correct so this PR migrates the DHT encryption to use the audited RustCrypto ChaCha20 encryption. 

Had to bump the toolchain version for this update so updated to the most current version of nightly. The only update needed was adding a new feature flag #![feature(min_type_alias_impl_trait)] which is a new flag for the same issue we already use (i.e. #![feature(type_alias_impl_trait)])

The new nightly toolchain contained an update to the linting used by rustfmt so I have included a second commit that is just new formatting to make this easier to review.

## How Has This Been Tested?
Tests updated

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [ ] I'm merging against the `development` branch.
* [ ] I have squashed my commits into a single commit.

CjS77 commented on comms/dht/src/crypt.rs:
```diff
@@ -20,29 +20,55 @@
 
-pub fn decrypt(cipher_key: &CommsPublicKey, cipher_text: &[u8]) -> Result<Vec<u8>, CipherError> {
-    ChaCha20::open_with_integral_nonce(&cipher_text.to_vec(), cipher_key.as_bytes())
+pub fn decrypt(cipher_key: &CommsPublicKey, cipher_text: &[u8]) -> Result<Vec<u8>, DhtOutboundError> {
```
Is decypting with a pubkey correct? How is the en- / decrpytion used in the comms? If it's all with pubkeys, then the encryption is likely meaningless.

philipr-za replied:
Its a Diffie-Hellman key

stringhandler replied:
The type here is confusing. Perhaps just a different type?

CjS77 replied:
DH keys are typically secret keys. Yeah, it's just confusing.

philipr-za replied:
Would need to update the way its implemented in the `tari_crypto` crate. 
https://github.com/tari-project/tari-crypto/blob/main/src/keys.rs#L78

CjS77 replied:
Yes, it's incorrect in tari_crypto. @hansieodendaal picked that up a few weeks ago and was going to push a PR changing it.

CjS77 commented on comms/dht/src/crypt.rs:
```diff
@@ -63,10 +89,10 @@ mod test {
         let key = CommsPublicKey::default();
         let cipher_text =
-            from_hex("7ecafb4c0a88325c984517fca1c529b3083e9976290a50c43ff90b2ccb361aeaabfaf680e744b96fc3649a447b")
+            from_hex("24bf9e698e14938e93c09e432274af7c143f8fb831f344f244ef02ca78a07ddc28b46fec536a0ca5c04737a604")
```
This is odd, I thought only edge cases were broken on the handrolled chacha. If so, then this should still work with the same input and this suggests that the API has changed significantly, meaning other parts of comms could break.

philipr-za replied:
So looked further into this and I cannot actually get the RustCrypto output to agree with the custom implementation or with the test vectors defined in https://datatracker.ietf.org/doc/html/rfc8439#section-2.4.1. However, I checked the output against the Python Cryptodome library and the RustCrypto and Cryptodome implementations agree 🤷‍♂️

CjS77 commented on comms/src/lib.rs:
```diff
@@ -17,6 +17,7 @@
 #![recursion_limit = "512"]
 // Allow `type Future = impl Future`
 #![feature(type_alias_impl_trait)]
+#![feature(min_type_alias_impl_trait)]
```
This feature [may not be stabilised for a while](https://github.com/rust-lang/rust/issues/63063#issuecomment-795682305) so you might be hampering the effort to move Tari to stable Rust.

philipr-za replied:
It is the same issue the feature above which is critical to large portions of the code, they share the same tracking Issue in the Rust codebase.

CjS77 reviewed (commented):
A quick overview. I have a few comments and questions.

philipr-za commented:
Updating the rust toolchain in this PR resulted in the rustfmt and clippy being updated so most of the PR are those updates. Just for clarity the only file that was actually updated in terms of logic is noted below.

philipr-za commented on comms/dht/src/crypt.rs:
```diff
@@ -20,29 +20,62 @@
-        },
-        ByteArray,
-    },
+    tari_utilities::ByteArray,
```
This is the file that was modified in this PR.

philipr-za committed:
Replace custom ChaCha20 cipher with RustCrypto ChaCha20 implementation

Fuzzing revealed that the custom ChaCha20 implementation in `tari_utilities` was not correct so this PR migrates the DHT encryption to use the audited RustCrypto ChaCha20 encryption. 

Had to bump the toolchain version for this update so updated to the most current version of nightly. The only update needed was adding a new feature flag #![feature(min_type_alias_impl_trait)] which is a new flag for the same issue we already use (i.e. #![feature(type_alias_impl_trait)])
